system-addon/content-src/components/Sections/SectionsPerfTimer.jsx

```diff
@@ -0,0 +1,104 @@
+const React = require("react");
+const {actionCreators: ac, actionTypes: at} = require("common/Actions.jsm");
+const {perfService: perfSvc} = require("common/PerfService.jsm");
+
+// Currently record only a fixed set of sections. This will prevent data
+// from custom sections from showing up or from topstories.
+const RECORDED_SECTIONS = ["highlights"];
+
+class SectionsPerfTimer extends React.Component {
+  constructor(props) {
+    super(props);
+    // Just for test dependency injection:
+    this.perfSvc = this.props.perfSvc || perfSvc;
+
+    this._sendBadStateEvent = this._sendBadStateEvent.bind(this);
+    this._reportMissingData = false;
+    this._timestampHandled = false;
+  }
+
+  componentDidMount() {
+    this._maybeSendPaintedEvent();
+  }
+
+  componentDidUpdate() {
+    this._maybeSendPaintedEvent();
+  }
+
+  /**
+   * Call the given callback after the upcoming frame paints.
+   *
+   * @note Both setTimeout and requestAnimationFrame are throttled when the page
+   * is hidden, so this callback may get called up to a second or so after the
+   * requestAnimationFrame "paint" for hidden tabs.
+   *
+   * Newtabs hidden while loading will presumably be fairly rare (other than
+   * preloaded tabs, which we will be filtering out on the server side), so such
+   * cases should get lost in the noise.
+   *
+   * If we decide that it's important to find out when something that's hidden
+   * has "painted", however, another option is to post a message to this window.
+   * That should happen even faster than setTimeout, and, at least as of this
+   * writing, it's not throttled in hidden windows in Firefox.
+   *
+   * @param {Function} callback
+   *
+   * @returns void
+   */
+  _afterFramePaint(callback) {
+    requestAnimationFrame(() => setTimeout(callback, 0));
+  }
+
+  _maybeSendPaintedEvent() {
+    if (RECORDED_SECTIONS.indexOf(this.props.id) === -1) {
+      return;
+    }
+
+    if (!this.props.initialized) {
+      // Remember to report back when data is available.
+      this._reportMissingData = true;
+    } else if (this._reportMissingData) {
+      const dataReadyKey = `${this.props.id}_data_ready_ts`;
+      this._reportMissingData = false;
+      // Report that data is available later than first render call.
+      this._afterFramePaint(() => this._sendBadStateEvent(dataReadyKey));
+    }
+
+    // Only record first call to render.
+    if (this._timestampHandled) {
+      return;
+    }
+
+    this._timestampHandled = true;
+
+    const firstPaintKey = `${this.props.id}_first_painted_ts`;
+    this._afterFramePaint(() => this.perfSvc.mark(firstPaintKey));
```

Comment:
I'm confused.  Doesn't this mean that topsites_first_painted_ts is going to be marked both in the Sections timer as well as in the Topsites timer?  Or is topsites not a Section?

Reply:
I see now that topsites is not in the RECORDED_SECTIONS array, but if it were, would it simplify the patch substantially and allow some code removal?  Or is that too much code churn for this moment?